src/main/thrift/parquet.thrift

```diff
@@ -191,6 +191,73 @@ enum FieldRepetitionType {
   REPEATED = 2;
 }
 
+/**
+  * A histogram of repetition and definition levels for either a page or column
+  * chunk.
+  *
+  * This is useful for:
+  *   1. Estimating the size of the data when materialized in memory
+  *
+  *   2. For filter push-down on nulls at various levels of nested
+  *   structures and list lengths.
+  */
+struct RepetitionDefinitionLevelHistogram {
+   /**
+    * When present, there is expected to be one element corresponding to each
+    * repetition (i.e. size=max repetition_level+1) where each element
+    * represents the number of times the repetition level was observed in the
+    * data.
+    *
+    * This field may be omitted if max_repetition_level is 0.
+    **/
+   1: optional list<i64> repetition_level_histogram;
+   /**
+    * Same as repetition_level_histogram except for definition levels.
+    *
+    * This field may be omitted if max_definition_level is 0 or 1.
+    **/
+   2: optional list<i64> definition_level_histogram;
+ }
+
+/**
+ * A structure for capturing metadata for estimating the unencoded,
+ * uncompressed size of data written. This is useful for readers to estimate
+ * how much memory is needed to reconstruct data in their memory model and for
+ * fine grained filter pushdown on nested structures (the histogram contained
+ * in this structure can help determine the number of nulls at a particular
+ * nesting level).
+ *
+ * Writers should populate all fields in this struct except for the exceptions
+ * listed per field.
+ */
+struct SizeStatistics {
+   /**
+    * The number of physical bytes stored for BYTE_ARRAY data values assuming
+    * no encoding. This is exclusive of the bytes needed to store the length of
+    * each byte array. In other words, this field is equivalent to the `(size
+    * of PLAIN-ENCODING the byte array values) - (4 bytes * number of values
+    * written)`. To determine unencoded sizes of other types readers can use
+    * schema information multiplied by the number of non-null and null values.
+    * The number of null/non-null values can be inferred from the histograms
+    * below.
+    *
+    * For example, if a column chunk is dictionary-encoded with dictionary
+    * ["a", "bc", "cde"], and a data page contains the indices [0, 0, 1, 2],
+    * then this value for that data page should be 7 (1 + 1 + 2 + 3).
+    *
+    * This field should only be set for types that use BYTE_ARRAY as their
+    * physical type.
+    */
+   1: optional i64 unencoded_byte_array_data_bytes;
```

Comment:
I don't really understand this field. Most readers, I'd hope, are not decoding a column chunk or even a page at a time, but are rather reading a number of rows at a time. This avoids potentially exploding memory on highly compressed data. It is unclear how this could be used by such a reader?

It also seems off to me that it assumes readers can't preserve dictionaries, something both arrow-cpp and arrow-rs are able to do...

Reply:
> Most readers, I'd hope, are not decoding a column chunk or even a page at a time, but are rather reading a number of rows at a time. This avoids potentially exploding memory on highly compressed data.

Do you mean parquet-rs does not decompress an entire page at once?

Reply:
> It also seems off to me that it assumes readers can't preserve dictionaries

Hmmm right maybe dictionary size should be take into account. It would be ok for read dictionary into Plain data(like Binary / String), but read into dictionary indices would require just dictionary raw-binary size. Should we just reserve some info for dictionary in `ColumnChunk` @emkornfield (Or this is not required)?

Reply:
> Do you mean parquet-rs does not decompress an entire page at once?

It does currently decompress a page at a time, but doesn't decode. It is a valid point that we ideally wouldn't decompress either at this granularity. This not only avoids blowing up memory, especially on dictionary-encoded data, but also allows for more sophisticated predicate pushdown.

I wrote up how it works - https://arrow.apache.org/blog/2022/12/26/querying-parquet-with-millisecond-latency/.

Taking a step back my point was not to say that what arrow-rs does is necessarily more correct, but to ask what the decoding methodology is where these statistics would be useful. Any form of predicate pushdown would effectively make these statistics a potentially generous over-estimate? Is there some particular problem this format addition is trying to solve?

Reply:
I agree that this field would not very useful for dictionary-encoded columns, but the field is optional anyway and readers are free to ignore it. 

If the byte array column is not dictionary-encoded, though (for example it's delta-encoded or simply plain-encoded), then this field would be useful to more efficiently presize the target data buffer for decoding.

Perhaps the wording "Writers should populate all fields in this struct" above should be relaxed? There is no need to mandate the fields' presence, IMHO.

Reply:
> would be useful to more efficiently presize the target data buffer for decoding

Only if you intend to read the entire page in one go? Are there readers that do this?

Reply:
> Only if you intend to read the entire page in one go? Are there readers that do this?

Hmm... good question, actually. I looked at the Parquet C++ decoders and none does that. Or, more precisely, the DELTA_BYTE_ARRAY decoder does decode all lengths up front, but the variable-length data is produced incrementally.

So I retract my explanation. Perhaps @emkornfield can chime in.

Reply:
Nice catch, I guess @etseidl would like to decompress-and-decode pre-allocate the page once, since they're on GPU?

Reply:
> Only if you intend to read the entire page in one go? Are there readers that do this?

@tustvold Yes, there are :) I am currently using a library called libcudf to process parquet files on a GPU. To make efficient use of the GPUs resources, you find yourself processing at the scale of thousands of pages at a time. As pages are decoded, the data is written to GPU RAM resident dataframes patterned on Arrow.  Due to the random order in which pages might be processed, the storage for these dataframes needs to be allocated up front, and the page data written to the correct location.  For fixed-width data this isn't such a heavy lift, but for byte arrays you need to know exactly how much data you'll have when fully materialized in order to do this. When I first started using the library, the smallest unit of work was the row group.  Needless to say it is easy for a row group to explode in size, rendering it impossible to fully materialize in the constrained environment of a GPU. A batched reader has recently been added to the library which will read row ranges that are known to fit in RAM, but as currently implemented it needs to do a pass through the data to get precise per-page sizing information to be able to figure out the row boundaries that will fit. For dictionary encoded byte arrays, it is necessary to fully decode the indices to find the length of each record in the page and sum these lengths to get the unencoded page size. For DELTA_BYTE_ARRAY one needs to decode the prefix and suffix lengths fully to get the same information. For PLAIN encoding, assuming you need to read the entire page you can simply subtract `4 * num_values` from the page size to get the materialized size, but to read a partial page you have to step through encoded lengths, skipping the inline byte data. In addition to this, I am using libcudf to perform compactions on sets of sorted parquet files that comprise a log structured merge tree. Here too, we cannot fit all of the files in GPU RAM for processing, so we have to each of the input files in stripes, where each stripe is a range on the keying column(s). Knowing the unencoded page sizes up front is a must to determine the most efficient key ranges to operate over. 

All this is to say that in my case, having the `unencoded_byte_array_data_bytes` field available at the page level is a life saver in terms of reducing the amount of preprocessing work that needs to be done to efficiently read parquet on the GPU. As an aside, I'll mention that the first thing I wanted to add to libcudf when I became aware of it was the ability to write the Column and Offset indexes, since my project makes heavy use of them. Paradoxically, when I tried to use the indexes to do page pruning, I found that while I could reduce the amount of file data read dramatically, my overall execution times did not decrease much...this was due to the fact that it takes roughly as much time to decode 1000 pages as it does a few dozen. Much of that is due to the fact that with a 20000 row limit on data page size, you can wind up with data pages in the 10s of kB range, but dictionaries in the megabytes. You can decompress a lot of small data pages in the time it takes to handle one large dictionary page.

Reply:
> I agree that this field would not very useful for dictionary-encoded columns, but the field is optional anyway and readers are free to ignore it.

Again, for my uses this field is _most_ useful for dictionary encoded columns which are the hardest to get sizing info for, and which lead to the greatest memory explosion upon decoding.

Reply:
I guess in another case, like we're now trying to decode Dictionary Encoded Column to in-memory dictionary (like arrow/arrow-rs and Velox does), this might be a bit tricky

Reply:
I think the main point touched on already by a few people but to made explicit is that there is no reader of parquet data that I'm aware of that keeps the parquet encoded format entirely for its memory model.  Some may support a few options here (e.g. as noted arrow supported dictionary and run-end encoding) but more generally there are many readers that transpose early to a row format (avro or parquet) and in those contexts having a final memory estimate can improve planning.   Even for readers that support dictionary encoding, it is not guaranteed that all pages will be dictionary encoded (i.e. dictionary grows too large and there is fallback).  In these cases being able to get a good estimate across all pages is useful.

Given no universal encoding knowing an upper bound on total size helps readers plan on a batch size to use in the face of memory memory pressure by calculating an estimated row size.

Even if there was an immediate reader that all parquet encodings that there isn't any massaging, often times data is transposed later in query processing to a row oriented format which often times will be "plain encoded".  For example for joins.  Having a good memory estimate here can help for join planning (especially for distributed joins).

Reply:
In which case might it be an idea to call this plain_encoded_size? This would perhaps make it more obvious what the meaning of the value is and why it is here?

Reply:
PLAIN encoding would include the value lengths, right? Here we are only talking about the variable-length data itself.

Reply:
Indeed, but one can convert between the two very easily and the advantage of PLAIN encoding is it is a well-defined reference point. IDK was just an idea, the whole notion of unencoded bytes seems non-obvious to me

Reply:
> PLAIN encoding would include the value lengths, right? Here we are only talking about the variable-length data itself.

yes, conversion isn't too hard.  In prior iterations of the CL I believe others had concern about PLAIN encoding being potentially harder to understand (I could be misremembering).  Plain encoding also seems like it would make implementation a little bit more awkward.  I imagine there will be a method like `UpdateStats(string x)`, as defined this is just `unencoded_bytes += x.size`, if it is plain encoding we have `unencoded_bytes += x.size + 4`, which again is minor but I think it adds noise?

Reply:
Personally, I think the current semantics as proposed by @emkornfield  make more sense. If you're decoding from a non-PLAIN encoding, it is unlikely that you are decoding _to_ a PLAIN encoding. Instead, you will probably decode to some internal format that's not Parquet, but perhaps or Arrow Arrow-similar. So knowing the data payload length is more immediately useful (for example it directly gives you the size of a binary column's data buffer in the Arrow format).